crates/ide-db/src/rename.rs

```diff
@@ -71,12 +71,25 @@ impl Definition {
         sema: &Semantics<'_, RootDatabase>,
         new_name: &str,
     ) -> Result<SourceChange> {
+        if let Some(krate) = self.krate(sema.db) {
+            if !matches!(krate.origin(sema.db), CrateOrigin::Local { .. }) {
+                bail!("Cannot rename a non-local definition.")
+            }
+        }
+
         match *self {
             Definition::Module(module) => rename_mod(sema, module, new_name),
             Definition::BuiltinType(_) => {
                 bail!("Cannot rename builtin type")
             }
             Definition::SelfType(_) => bail!("Cannot rename `Self`"),
+            Definition::Macro(mac) => {
+                if mac.is_builtin_derive(sema.db) {
+                    bail!("Cannot rename builtin macro")
+                }
```

Comment:
Can you also add a test for this?

Reply:
I can't seem to find a way that would bypass `Cannot rename a non-local definition.` and bail out at `Cannot rename builtin derive`